### PR TITLE
add filename_as_doc_id

### PR DESCRIPTION
**Pull request by logan-markewich**
# Description

The `refresh()` function would be a lot easier to use if `doc_id()` was more consistent and automatic. This PR adds an option to have the `doc_id` be set to the filepath

Fixes https://github.com/jerryjliu/llama_index/issues/5772

(Also threw in some extra docs fixes for discover llamaindex)

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Added new unit/integration tests
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Disiok — APPROVED** [review]
Great feature for better usability

**Review by Disiok — APPROVED** [review]
Great feature for better usability

File: llama_index/readers/file/base.py [hunk 1]
```diff
@@ -188,13 +190,22 @@ def load_data(self) -> List[Document]:
                     self.file_extractor[file_suffix] = reader_cls()
                 reader = self.file_extractor[file_suffix]
                 docs = reader.load_data(input_file, extra_info=metadata)
+
+                # iterate over docs if needed
+                if self.filename_as_id:
+                    for i in range(len(docs)):
```
Comment (Disiok) [inline llama_index/readers/file/base.py:196]: nit: why not just `for doc in docs`?

File: llama_index/readers/file/base.py [hunk 2]
```diff
@@ -188,13 +190,22 @@ def load_data(self) -> List[Document]:
...
+                        docs[i].doc_id = f"{str(Path(input_file))}_part_{i}"
```
Comment (Disiok) [inline llama_index/readers/file/base.py:197]: very nit: isn't `input_file` already a Path?

File: llama_index/readers/file/base.py (hunk 1 above)
Comment (logan-markewich) [reply llama_index/readers/file/base.py]: Good point

File: llama_index/readers/file/base.py (hunk 2 above)
Comment (logan-markewich) [reply llama_index/readers/file/base.py]: yea, whoops

**Pull request merged**